ariadne/cardano/src/Ariadne/MainTemplate.hs

```diff
@@ -163,10 +167,24 @@ initializeEverything MainSettings {..}
       -- Spawn backend actions in async thread, then run ui action in the main thread
       -- This is needed because some UI libraries (Qt) insist on livng in the main thread
       withAsync serviceAction $ \serviceThread -> do
-        -- Make backend rethrow all exceptions in main thread if something goes wrong
-        link serviceThread
-        -- TODO (AD-432) if uiAction blocks exceptions, for example by entering FFI call and not
-        -- returning from it, exceptions from the linked thread won't be rethrown correctly.
+        -- Make custom link to service thread for UI apps.
+        -- It is going to call msKillUI and rethrow exception, if somethings goes wrong
+        linkUI serviceThread msKillUI
         uiAction
 
   return mainAction
+
+-- Similar to link from Control.Concurrent.Async, but calls finalizer on exception
+linkUI :: Async a -> (SomeException -> IO ()) -> IO ()
+linkUI thread finalizer = do
+  me <- myThreadId
+  void $ forkIO $ do
+    r <- waitCatch thread
+    case r of
+      Left e | (not . isCancel) e -> finalizer e >> throwTo me (ExceptionInLinkedThread thread e)
```

Comment:
Remind me, what's with "what if finalizer throws" question?

Reply:
Also, strange brackets :)

Reply:
Can you justify using `forkIO` instead of `forkRepeat`?

Reply:
AFAIU it will catch the exception and then finalize and rethrow it, while fork repeat will also re-run this action

Reply:
And exception will be correcrtly rethrown, correct me, if I'm wrong